core/controllers/diagnostic_test_player.py

```diff
@@ -77,30 +104,52 @@ def get(self, topic_id: str) -> None:
 
         diagnostic_test_skill_ids = topic.skill_ids_for_diagnostic_test
 
-        # From each skill, two questions were fetched. The first question
-        # (main question) will be presented to the learner initially and if the
-        # learner attempted incorrectly, then another question from the same
-        # skill (backup question) will be presented otherwise not. Hence the
-        # number of questions is twice the length of skill IDs.
-        question_count = len(diagnostic_test_skill_ids) * 2
-
+        skill_id_to_all_questions_fetchable_at_one_time = (
+            collections.defaultdict(list))
+        question_ids_for_fetched_skills = []
+
+        for skill_id in diagnostic_test_skill_ids:
+            questions = question_services.get_questions_by_skill_ids(
+                feconf.MAX_QUESTIONS_FETCHABLE_AT_ONE_TIME,
+                [skill_id],
+                require_medium_difficulty=True
+            )
+            skill_id_to_all_questions_fetchable_at_one_time[skill_id] = (
+                questions)
+
+            question_ids_for_fetched_skills.extend(
+                [question.id for question in questions])
+
+        # A dict with skill ID as key and a list with a maximum of two
+        # questions as value. Among the two questions, one question should be
+        # considered as the main and the other should be considered as
+        # the backup.
         skill_id_to_questions_map: Dict[
             str, List[question_domain.Question]] = (
                 collections.defaultdict(list))
 
-        questions = question_services.get_questions_by_skill_ids(
-            question_count,
-            diagnostic_test_skill_ids,
-            require_medium_difficulty=True
-        )
-
-        for question in questions:
-            linked_skill_ids = question.linked_skill_ids
-            diagnostic_test_linked_skill_ids = list(
-                set(linked_skill_ids) & set(diagnostic_test_skill_ids))
-
-            for skill_id in diagnostic_test_linked_skill_ids:
-                skill_id_to_questions_map[skill_id].append(question)
+        for skill_id, questions in (
+            skill_id_to_all_questions_fetchable_at_one_time.items()):
+            question_ids_for_other_diagnostic_test_skills = list(
+                set(question_ids_for_fetched_skills) -
+                set(question.id for question in questions)
+            )
+            # A list of question IDs, where each ID belongs to the questions
+            # that are already presented in the test or that belong to other
+            # skills. These IDs should not be considered in order to avoid the
+            # repetition of questions in the diagnostic test.
+            question_ids_to_exclude = (
+                question_ids_for_other_diagnostic_test_skills +
+                exclude_question_ids
```

Comment:
Perhaps excluded_question_ids is better.

Reply:
Thank you for the suggestion, this sounds clearer.
I have updated the var name.